README.md
[![Software License](https://img.shields.io/badge/License-MIT-orange.svg?style=flat-square)](https://github.com/etalab/jours-feries-france/blob/master/LICENSE.md)
![CircleCI](https://img.shields.io/circleci/project/github/etalab/jours-feries-france.svg?style=flat-square)
![PyPI - Downloads](https://img.shields.io/pypi/dm/jours-feries-france.svg?style=flat-square)

# Jours fériés France
Cette librairie calcule les jours fériés en France. Vous pouvez découvrir plus de détails sur les jours fériés français sur [service-public.fr](https://www.service-public.fr/particuliers/vosdroits/F2405).

Cette librairie fonctionne à partir de la version 2.7 de Python.

## Installation
```
pip install jours-feries-france
```

## Usage
```python
import datetime

from jours_feries_france import JoursFeries

# Obtenir les jours fériés pour une année, pour la métropole
res = JoursFeries.for_year(2018)
# res est un dictionnaire
# {
#     '1er janvier': datetime.date(2018, 1, 1),
#     'Lundi de Pâques': datetime.date(2018, 4, 2),
#     '1er mai': datetime.date(2018, 5, 1),
#     '8 mai': datetime.date(2018, 5, 8),
#     'Ascension': datetime.date(2018, 5, 10),
#     'Lundi de Pentecôte': datetime.date(2018, 5, 21),
#     '14 juillet': datetime.date(2018, 7, 14),
#     'Assomption': datetime.date(2018, 8, 15),
#     'Toussaint': datetime.date(2018, 11, 1),
#     '11 novembre': datetime.date(2018, 11, 11),
#     'Jour de Noël': datetime.date(2018, 12, 25)
# }

# Vous pouvez aussi obtenir certains jours fériés en tant que datetime.date
print (JoursFeries.lundi_paques(2018))
print (JoursFeries.ascension(2018))
print (JoursFeries.lundi_pentecote(2018))

# Obtenir les jours fériés pour une zone spécifique
res = JoursFeries.for_year(2018, zone="Alsace-Moselle")

# Quelques fonctions d'aide
JoursFeries.is_bank_holiday(datetime.date(2019, 12, 25), zone="Métropole")
# -> True
JoursFeries.next_bank_holiday(datetime.date(2019, 12, 24), zone="Métropole")
# -> ('Noël', datetime.date(2019, 12, 25))
```

### Zones disponibles
Les zones suivantes sont disponibles :
- `Métropole` (par défaut)
- `Alsace-Moselle`
- `Guadeloupe`
- `Guyane`
- `La Réunion`
- `Martinique`
- `Mayotte`
- `Nouvelle-Calédonie`
- `Polynésie Française`
- `Saint-Barthélémy`
- `Saint-Martin`
- `Saint-Pierre-et-Miquelon`
- `Wallis-et-Futuna`

Ces zones sont disponibles dans une constante :
```python
from jours_feries_france import JoursFeries

print(JoursFeries.ZONES)
# [
#     'Métropole', 'Alsace-Moselle', 'Guadeloupe',
#     'Guyane', 'Martinique', 'Mayotte', 'Nouvelle-Calédonie',
#     'La Réunion', 'Polynésie Française', 'Saint-Barthélémy',
#     'Saint-Martin', 'Wallis-et-Futuna', 'Saint-Pierre-et-Miquelon'
# ]
```

### Noms des jours fériés
Les noms des jours fériés sont fixés d'[après le code du travail](#sources).

Pour les zones `Métropole` (par défaut), `Nouvelle-Calédonie`, `Polynésie Française`, `Saint-Pierre-et-Miquelon` et `Wallis-et-Futuna`, les jours fériés sont orthographiés de la façon suivante :
- `1er janvier`
- `Lundi de Pâques`
- `1er mai`
- `8 mai`
- `Ascension`
- `Lundi de Pentecôte`
- `14 juillet`
- `Assomption`
- `Toussaint`
- `11 novembre`
- `Jour de Noël`

Pour la zone `Alsace-Moselle`, il existe 2 jours fériés supplémentaires, orthographiés de la façon suivante :
- `2ème jour de Noël`
- `Vendredi saint`

Pour les zones `Guadeloupe`, `Guyane`, `La Réunion`, `Martinique`, `Mayotte`, `Saint-Barthélémy` et `Saint-Martin`, il existe un jour férié supplémentaire, orthographié de la façon suivante :
- `Abolition de l'esclavage`

### Noms des méthodes Python

Vous pouvez calculer chaque jour férié individuellement à l'aide d'une méthode spécifique.

```python
from jours_feries_france import JoursFeries

year = 2020
zone = 'Métropole'

print("1er janvier", JoursFeries.premier_janvier(year))
print("1er mai", JoursFeries.premier_mai(year))
print("8 mai", JoursFeries.huit_mai(year))
print("14 juillet", JoursFeries.quatorze_juillet(year))
print("Assomption", JoursFeries.assomption(year))
print("Toussaint", JoursFeries.toussaint(year))
print("11 novembre", JoursFeries.onze_novembre(year))
print("Jour de Noël", JoursFeries.jour_noel(year))
print("Lundi de Pâques", JoursFeries.lundi_paques(year))
print("Ascension", JoursFeries.ascension(year))
print("Lundi de Pentecôte", JoursFeries.lundi_pentecote(year))
print("Vendredi saint", JoursFeries.vendredi_saint(year, zone))
print("2ème jour de Noël", JoursFeries.deuxieme_jour_noel(year, zone))
print("Abolition de l'esclavage", JoursFeries.abolition_esclavage(year, zone))
```

Certaines méthodes acceptent une `zone` en paramètre car ce jour férié est spécifique à certaines zones. Si ce jour férié n'est pas férié pour la zone passée en argument, vous aurez la valeur `None` en retour au lieu d'une date.

## Données
Si vous souhaitez simplement un export, consultez le jeu de données ["Jours fériés en France"](https://www.data.gouv.fr/fr/datasets/jours-feries-en-france/) sur data.gouv.fr.

## Sources
La liste des jours fériés est définie [dans le code du travail](#sources).

Certaines commémorations locales ou professionnelles sont également des jours fériés, parmi lesquelles :
- Saint-Éloi (reconnu jour férié par certaines conventions collectives dans la métallurgie) ;
- Sainte-Barbe (pour les salariés travaillant dans les mines) ;
- Mi-carême dans certains DOM.

Ces fêtes locales ou professionnelles ne sont pas disponibles dans cette librairie.

- [Code du travail : articles L3133-1 à L3133-3](https://www.legifrance.gouv.fr/affichCode.do?idSectionTA=LEGISCTA000033008129&cidTexte=LEGITEXT000006072050)
- [Code du travail - Article L3422-2](https://www.legifrance.gouv.fr/affichCodeArticle.do?idArticle=LEGIARTI000035902463&cidTexte=LEGITEXT000006072050)


CHANGELOG.md
# Changements

## 0.7.0 - 2020-05-15

- ⚠️ Utilisation des noms des jours fériés en vigueur dans le code du travail
- ⚠️ Changement du nom des méthodes pour chaque jour férié pour correspondre au code du travail


tests/test.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

import unittest
from datetime import date

from jours_feries_france import JoursFeries


class TestDatasetParser(unittest.TestCase):
    def test_validates_region(self):
        with self.assertRaises(ValueError):
            JoursFeries.for_year(2018, zone="foo")

    def test_is_bank_holiday(self):
        self.assertTrue(JoursFeries.is_bank_holiday(date(2019, 12, 25)))
        self.assertTrue(
            JoursFeries.is_bank_holiday(date(2019, 12, 25), zone="Métropole")
        )
        self.assertTrue(
            JoursFeries.is_bank_holiday(date(2019, 12, 26), zone="Alsace-Moselle")
        )

        self.assertFalse(JoursFeries.is_bank_holiday(date(2019, 12, 26)))
        self.assertFalse(
            JoursFeries.is_bank_holiday(date(2019, 12, 26), zone="Métropole")
        )

    def test_next_bank_holiday(self):
        self.assertEquals(
            ("11 novembre", date(2018, 11, 11)),
            JoursFeries.next_bank_holiday(date(2018, 11, 10)),
        )

        self.assertEquals(
            ("11 novembre", date(2018, 11, 11)),
            JoursFeries.next_bank_holiday(date(2018, 11, 11), zone="Métropole"),
        )

        self.assertEquals(
            ("Jour de Noël", date(2018, 12, 25)),
            JoursFeries.next_bank_holiday(date(2018, 12, 11), zone="Métropole"),
        )

    def test_for_year(self):
        self.assertDictEqual(
            JoursFeries.for_year(2018),
            {
                "1er janvier": date(2018, 1, 1),
                "Lundi de Pâques": date(2018, 4, 2),
                "1er mai": date(2018, 5, 1),
                "8 mai": date(2018, 5, 8),
                "Ascension": date(2018, 5, 10),
                "Lundi de Pentecôte": date(2018, 5, 21),
                "14 juillet": date(2018, 7, 14),
                "Assomption": date(2018, 8, 15),
                "Toussaint": date(2018, 11, 1),
                "11 novembre": date(2018, 11, 11),
                "Jour de Noël": date(2018, 12, 25),
            },
        )

        self.assertDictEqual(
            JoursFeries.for_year(2020),
            {
                "11 novembre": date(2020, 11, 11),
                "Ascension": date(2020, 5, 21),
                "Assomption": date(2020, 8, 15),
                "14 juillet": date(2020, 7, 14),
                "1er mai": date(2020, 5, 1),
                "1er janvier": date(2020, 1, 1),
                "Lundi de Pâques": date(2020, 4, 13),
                "Jour de Noël": date(2020, 12, 25),
                "Lundi de Pentecôte": date(2020, 6, 1),
                "Toussaint": date(2020, 11, 1),
                "8 mai": date(2020, 5, 8),
            },
        )

    def test_for_year_in_alsace(self):
        self.assertDictEqual(
            JoursFeries.for_year(2018, zone="Alsace-Moselle"),
            {
                "11 novembre": date(2018, 11, 11),
                "Ascension": date(2018, 5, 10),
                "Assomption": date(2018, 8, 15),
                "14 juillet": date(2018, 7, 14),
                "1er mai": date(2018, 5, 1),
                "1er janvier": date(2018, 1, 1),
                "Lundi de Pâques": date(2018, 4, 2),
                "Jour de Noël": date(2018, 12, 25),
                "Lundi de Pentecôte": date(2018, 5, 21),
                "Toussaint": date(2018, 11, 1),
                "8 mai": date(2018, 5, 8),
                "Vendredi saint": date(2018, 3, 30),
                "2ème jour de Noël": date(2018, 12, 26),
            },
        )

        self.assertDictEqual(
            JoursFeries.for_year(2020, zone="Alsace-Moselle"),
            {
                "11 novembre": date(2020, 11, 11),
                "Ascension": date(2020, 5, 21),
                "Assomption": date(2020, 8, 15),
                "14 juillet": date(2020, 7, 14),
                "1er mai": date(2020, 5, 1),
                "1er janvier": date(2020, 1, 1),
                "Lundi de Pâques": date(2020, 4, 13),
                "Jour de Noël": date(2020, 12, 25),
                "Lundi de Pentecôte": date(2020, 6, 1),
                "Toussaint": date(2020, 11, 1),
                "8 mai": date(2020, 5, 8),
                "Vendredi saint": date(2020, 4, 10),
                "2ème jour de Noël": date(2020, 12, 26),
            },
        )

    def testPaques(self):
        self.assertEquals(JoursFeries.paques(1954), date(1954, 4, 18))
        self.assertEquals(JoursFeries.paques(1981), date(1981, 4, 19))
        self.assertEquals(JoursFeries.paques(2049), date(2049, 4, 18))

    def testNamesAllZones(self):
        def names(holidays):
            return set(holidays.keys())

        base = set(
            [
                "1er janvier",
                "1er mai",
                "8 mai",
                "14 juillet",
                "Assomption",
                "Toussaint",
                "11 novembre",
                "Jour de Noël",
                "Lundi de Pâques",
                "Ascension",
                "Lundi de Pentecôte",
            ]
        )

        extra_holidays = [
            ["Alsace-Moselle", set(["Vendredi saint", "2ème jour de Noël"])],
            ["Guadeloupe", set(["Abolition de l'esclavage"])],
            ["Guyane", set(["Abolition de l'esclavage"])],
            ["Martinique", set(["Abolition de l'esclavage"])],
            ["Mayotte", set(["Abolition de l'esclavage"])],
            ["Nouvelle-Calédonie", set()],
            ["La Réunion", set(["Abolition de l'esclavage"])],
            ["Polynésie Française", set()],
            ["Saint-Barthélémy", set(["Abolition de l'esclavage"])],
            ["Saint-Martin", set(["Abolition de l'esclavage"])],
            ["Wallis-et-Futuna", set()],
            ["Saint-Pierre-et-Miquelon", set()],
        ]

        self.assertSetEqual(names(JoursFeries.for_year(2020, "Métropole")), base)

        for test in extra_holidays:
            zone, holidays = test
            self.assertEquals(
                names(JoursFeries.for_year(2020, zone)), base.union(holidays)
            )

        self.assertEquals(
            JoursFeries.ZONES, ["Métropole"] + [e[0] for e in extra_holidays]
        )

    def testAbolition_esclavage(self):
        tests = [
            ("Mayotte", date(2020, 4, 27)),
            ("Martinique", date(2020, 5, 22)),
            ("Guadeloupe", date(2020, 5, 27)),
            ("Saint-Martin", date(2020, 5, 28)),
            ("Guyane", date(2020, 6, 10)),
            ("Saint-Barthélémy", date(2020, 10, 9)),
            ("La Réunion", date(2020, 12, 20)),
        ]

        zones = set()
        for test in tests:
            zone, expected_date = test
            zones.add(zone)
            self.assertEquals(
                JoursFeries.abolition_esclavage(2020, zone), expected_date
            )

        for zone in [z for z in JoursFeries.ZONES if z not in zones]:
            self.assertEquals(JoursFeries.abolition_esclavage(2020, zone), None)

        # Saint-Martin
        self.assertEquals(
            JoursFeries.abolition_esclavage(2017, "Saint-Martin"), date(2017, 5, 27)
        )
        self.assertEquals(
            JoursFeries.abolition_esclavage(2018, "Saint-Martin"), date(2018, 5, 28)
        )


tests/test_compute.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

import unittest
from datetime import date

from jours_feries_france.compute import JoursFeries


class TestDatasetParser(unittest.TestCase):
    def test_validates_region(self):
        with self.assertRaises(ValueError):
            JoursFeries.for_year(2018, zone="foo")

    def test_is_holiday(self):
        self.assertTrue(JoursFeries.is_holiday(date(2019, 12, 25)))
        self.assertTrue(JoursFeries.is_holiday(date(2019, 12, 25), zone="Métropole"))
        self.assertTrue(
            JoursFeries.is_holiday(date(2019, 12, 26), zone="Alsace-Moselle")
        )

        self.assertFalse(JoursFeries.is_holiday(date(2019, 12, 26)))
        self.assertFalse(JoursFeries.is_holiday(date(2019, 12, 26), zone="Métropole"))

    def test_next_holiday(self):
        self.assertEquals(
            ("Armistice", date(2018, 11, 11)),
            JoursFeries.next_holiday(date(2018, 11, 10)),
        )

        self.assertEquals(
            ("Armistice", date(2018, 11, 11)),
            JoursFeries.next_holiday(date(2018, 11, 11), zone="Métropole"),
        )

        self.assertEquals(
            ("Noël", date(2018, 12, 25)),
            JoursFeries.next_holiday(date(2018, 12, 11), zone="Métropole"),
        )

    def test_for_year(self):
        self.assertDictEqual(
            JoursFeries.for_year(2018),
            {
                "Jour de l'an": date(2018, 1, 1),
                "Lundi de Pâques": date(2018, 4, 2),
                "Fête du travail": date(2018, 5, 1),
                "Victoire des alliés": date(2018, 5, 8),
                "Ascension": date(2018, 5, 10),
                "Lundi de Pentecôte": date(2018, 5, 21),
                "Fête Nationale": date(2018, 7, 14),
                "Assomption": date(2018, 8, 15),
                "Toussaint": date(2018, 11, 1),
                "Armistice": date(2018, 11, 11),
                "Noël": date(2018, 12, 25),
            },
        )

        self.assertDictEqual(
            JoursFeries.for_year(2020),
            {
                "Armistice": date(2020, 11, 11),
                "Ascension": date(2020, 5, 21),
                "Assomption": date(2020, 8, 15),
                "Fête Nationale": date(2020, 7, 14),
                "Fête du travail": date(2020, 5, 1),
                "Jour de l'an": date(2020, 1, 1),
                "Lundi de Pâques": date(2020, 4, 13),
                "Noël": date(2020, 12, 25),
                "Lundi de Pentecôte": date(2020, 6, 1),
                "Toussaint": date(2020, 11, 1),
                "Victoire des alliés": date(2020, 5, 8),
            },
        )

    def test_for_year_in_alsace(self):
        self.assertDictEqual(
            JoursFeries.for_year(2018, zone="Alsace-Moselle"),
            {
                "Armistice": date(2018, 11, 11),
                "Ascension": date(2018, 5, 10),
                "Assomption": date(2018, 8, 15),
                "Fête Nationale": date(2018, 7, 14),
                "Fête du travail": date(2018, 5, 1),
                "Jour de l'an": date(2018, 1, 1),
                "Lundi de Pâques": date(2018, 4, 2),
                "Noël": date(2018, 12, 25),
                "Lundi de Pentecôte": date(2018, 5, 21),
                "Toussaint": date(2018, 11, 1),
                "Victoire des alliés": date(2018, 5, 8),
                "Vendredi Saint": date(2018, 3, 30),
                "Saint Étienne": date(2018, 12, 26),
            },
        )

        self.assertDictEqual(
            JoursFeries.for_year(2020, zone="Alsace-Moselle"),
            {
                "Armistice": date(2020, 11, 11),
                "Ascension": date(2020, 5, 21),
                "Assomption": date(2020, 8, 15),
                "Fête Nationale": date(2020, 7, 14),
                "Fête du travail": date(2020, 5, 1),
                "Jour de l'an": date(2020, 1, 1),
                "Lundi de Pâques": date(2020, 4, 13),
                "Noël": date(2020, 12, 25),
                "Lundi de Pentecôte": date(2020, 6, 1),
                "Toussaint": date(2020, 11, 1),
                "Victoire des alliés": date(2020, 5, 8),
                "Vendredi Saint": date(2020, 4, 10),
                "Saint Étienne": date(2020, 12, 26),
            },
        )

    def testPaques(self):
        self.assertEquals(JoursFeries.paques(1954), date(1954, 4, 18))
        self.assertEquals(JoursFeries.paques(1981), date(1981, 4, 19))
        self.assertEquals(JoursFeries.paques(2049), date(2049, 4, 18))

    def testNamesAllZones(self):
        def names(holidays):
            return set(holidays.keys())

        base = set(
            [
                "Jour de l'an",
                "Fête du travail",
                "Victoire des alliés",
                "Fête Nationale",
                "Assomption",
                "Toussaint",
                "Armistice",
                "Noël",
                "Lundi de Pâques",
                "Ascension",
                "Lundi de Pentecôte",
            ]
        )

        extra_holidays = [
            ["Alsace-Moselle", set(["Vendredi Saint", "Saint Étienne"])],
            ["Guadeloupe", set(["Abolition de l'esclavage"]),],
            ["Guyane", set(["Abolition de l'esclavage"])],
            ["Martinique", set(["Abolition de l'esclavage"]),],
            ["Mayotte", set(["Abolition de l'esclavage"]),],
            ["Nouvelle-Calédonie", set()],
            ["La Réunion", set(["Abolition de l'esclavage"])],
            ["Polynésie Française", set()],
            ["Saint-Barthélémy", set(["Abolition de l'esclavage"])],
            ["Saint-Martin", set(["Abolition de l'esclavage"])],
            ["Wallis-et-Futuna", set()],
        ]

        self.assertSetEqual(names(JoursFeries.for_year(2020, "Métropole")), base)

        for test in extra_holidays:
            zone, holidays = test
            self.assertEquals(
                names(JoursFeries.for_year(2020, zone)), base.union(holidays)
            )

        self.assertEquals(
            JoursFeries.ZONES, ["Métropole"] + [e[0] for e in extra_holidays]
        )

    def testAbolitionDeLesclavage(self):
        tests = [
            ("Mayotte", date(2020, 4, 27)),
            ("Martinique", date(2020, 5, 22)),
            ("Guadeloupe", date(2020, 5, 27)),
            ("Saint-Martin", date(2020, 5, 28)),
            ("Guyane", date(2020, 6, 10)),
            ("Saint-Barthélémy", date(2020, 10, 9)),
            ("La Réunion", date(2020, 12, 20)),
        ]

        zones = set()
        for test in tests:
            zone, expected_date = test
            zones.add(zone)
            self.assertEquals(
                JoursFeries.abolitionDeLesclavage(2020, zone), expected_date
            )

        for zone in [z for z in JoursFeries.ZONES if z not in zones]:
            self.assertEquals(JoursFeries.abolitionDeLesclavage(2020, zone), None)

        # Saint-Martin
        self.assertEquals(
            JoursFeries.abolitionDeLesclavage(2017, "Saint-Martin"), date(2017, 5, 27)
        )
        self.assertEquals(
            JoursFeries.abolitionDeLesclavage(2018, "Saint-Martin"), date(2018, 5, 28)
        )


main.py
import os
import json
import datetime
from collections import defaultdict

import pandas as pd
from ics import Calendar, Event
from jours_feries_france import JoursFeries
from slugify import slugify


def to_csv(df, filename):
    df.sort_values(by="date").to_csv(filename, index=False, encoding="utf-8")


def add_event(calendar, name, date):
    event = Event()
    event.name = name
    event.begin = date.strftime("%Y-%m-%d")
    event.make_all_day()
    calendar.events.add(event)


def write_calendar(calendar, filename, name):
    content = str(calendar).split("\n")

    # Add calendar name
    content.insert(2, f"NAME:{name}\r")
    content.insert(2, f"X-WR-CALNAME:{name}\r")

    with open(filename, "w") as f:
        f.writelines(content)


def write_json(filename, bank_holidays):
    with open(filename, "w") as f:
        json.dump(
            {k.strftime("%Y-%m-%d"): v for k, v in bank_holidays.items()},
            f,
            ensure_ascii=False,
        )


current_year = datetime.date.today().year
START, END = -20, 5

os.makedirs("data/csv", exist_ok=True)
os.makedirs("data/ics", exist_ok=True)
all_dates = defaultdict(list)
for zone, zone_slug in [(z, slugify(z)) for z in JoursFeries.ZONES]:
    os.makedirs(f"data/json/{zone_slug}", exist_ok=True)

    csv_data = []
    json_data = {}
    calendar = Calendar()
    calendar.creator = "Etalab"
    for year in range(current_year + START, current_year + END + 1):
        bank_holidays = JoursFeries.for_year(year, zone)

        for nom_jour_ferie, the_date in bank_holidays.items():
            # Generate ICS calendar only from 5 years ago to 5 years in the future
            is_recent = the_date.year in range(current_year - 5, current_year + 6)
            if is_recent:
                add_event(calendar, nom_jour_ferie, the_date)

            csv_data.append(
                {
                    "date": the_date.strftime("%Y-%m-%d"),
                    "annee": str(the_date.year),
                    "zone": zone,
                    "nom_jour_ferie": nom_jour_ferie,
                }
            )
            all_dates[(the_date.strftime("%Y-%m-%d"), nom_jour_ferie)].append(zone)

        json_for_year = {v: k for k, v in bank_holidays.items()}
        write_json(f"data/json/{zone_slug}/{year}.json", json_for_year)
        json_data = {**json_for_year, **json_data}

    to_csv(pd.DataFrame(csv_data), f"data/csv/jours_feries_{zone_slug}.csv")
    write_json(f"data/json/{zone_slug}.json", json_data)

    write_calendar(
        calendar, f"data/ics/jours_feries_{zone_slug}.ics", f"Jours fériés {zone}"
    )

res = []
for (date, name), zones in all_dates.items():
    res.append({"date": date, "nom_jour_ferie": name, "zones": "|".join(zones)})
to_csv(pd.DataFrame(res), f"data/csv/jours_feries.csv")


jours_feries_france/compute.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from datetime import date, timedelta


class JoursFeries(object):
    ZONES = [
        "Métropole",
        "Alsace-Moselle",
        "Guadeloupe",
        "Guyane",
        "Martinique",
        "Mayotte",
        "Nouvelle-Calédonie",
        "La Réunion",
        "Polynésie Française",
        "Saint-Barthélémy",
        "Saint-Martin",
        "Wallis-et-Futuna",
    ]

    def __init__(self):
        super(JoursFeries, self).__init__()

    @staticmethod
    def check_zone(zone):
        zone = zone or "Métropole"

        if zone not in JoursFeries.ZONES:
            valid_values = ", ".join(JoursFeries.ZONES)
            raise ValueError(f"{zone} is invalid. Supported values: {valid_values}")

        return zone

    @staticmethod
    def is_holiday(date, zone=None):
        return date in JoursFeries.for_year(date.year, zone).values()

    @staticmethod
    def next_holiday(date, zone=None):
        while not JoursFeries.is_holiday(date, zone):
            date = date + timedelta(days=1)

        return [
            (k, v)
            for k, v in JoursFeries.for_year(date.year, zone).items()
            if v == date
        ][0]

    @staticmethod
    def for_year(year, zone=None):
        JoursFeries.check_zone(zone)

        holidays = {
            "Jour de l'an": JoursFeries.jourDeLAn(year),
            "Fête du travail": JoursFeries.feteDuTravail(year),
            "Victoire des alliés": JoursFeries.victoireDesAllies(year),
            "Fête Nationale": JoursFeries.feteNationale(year),
            "Assomption": JoursFeries.assomption(year),
            "Toussaint": JoursFeries.toussaint(year),
            "Armistice": JoursFeries.armistice(year),
            "Noël": JoursFeries.noel(year),
            "Lundi de Pâques": JoursFeries.lundiDePaques(year),
            "Ascension": JoursFeries.ascension(year),
            "Lundi de Pentecôte": JoursFeries.lundiDePentecote(year),
            "Vendredi Saint": JoursFeries.vendrediSaint(year, zone),
            "Saint Étienne": JoursFeries.saintEtienne(year, zone),
            "Abolition de l'esclavage": JoursFeries.abolitionDeLesclavage(year, zone),
        }

        holidays = {k: v for k, v in holidays.items() if v}

        return {k: v for k, v in sorted(holidays.items(), key=lambda item: item[1])}

    @staticmethod
    def paques(year):
        if year < 1886:
            return None
        a = year % 19
        b = year // 100
        c = year % 100
        d = (19 * a + b - b // 4 - ((b - (b + 8) // 25 + 1) // 3) + 15) % 30
        e = (32 + 2 * (b % 4) + 2 * (c // 4) - d - (c % 4)) % 7
        f = d + e - 7 * ((a + 11 * d + 22 * e) // 451) + 114
        month = f // 31
        day = f % 31 + 1
        return date(year, month, day)

    @staticmethod
    def lundiDePaques(year):
        if year >= 1886:
            return JoursFeries.paques(year) + timedelta(days=1)
        return None

    @staticmethod
    def vendrediSaint(year, zone):
        if zone == JoursFeries.check_zone("Alsace-Moselle"):
            return JoursFeries.paques(year) - timedelta(days=2)
        return None

    @staticmethod
    def ascension(year):
        if year >= 1802:
            return JoursFeries.paques(year) + timedelta(days=39)
        return None

    @staticmethod
    def lundiDePentecote(year):
        if year >= 1886:
            return JoursFeries.paques(year) + timedelta(days=50)
        return None

    @staticmethod
    def jourDeLAn(year):
        if year > 1810:
            return date(year, 1, 1)
        return None

    @staticmethod
    def feteDuTravail(year):
        if year > 1919:
            return date(year, 5, 1)
        return None

    @staticmethod
    def victoireDesAllies(year):
        if (1953 <= year <= 1959) or year > 1981:
            return date(year, 5, 8)
        return None

    @staticmethod
    def feteNationale(year):
        if year >= 1880:
            return date(year, 7, 14)
        return None

    @staticmethod
    def toussaint(year):
        if year >= 1802:
            return date(year, 11, 1)
        return None

    @staticmethod
    def assomption(year):
        if year >= 1802:
            return date(year, 8, 15)
        return None

    @staticmethod
    def armistice(year):
        if year >= 1918:
            return date(year, 11, 11)
        return None

    @staticmethod
    def noel(year):
        if year >= 1802:
            return date(year, 12, 25)
        return None

    @staticmethod
    def saintEtienne(year, zone):
        if zone == JoursFeries.check_zone("Alsace-Moselle"):
            return date(year, 12, 26)
        return None

    @staticmethod
    def abolitionDeLesclavage(year, zone):
        if zone == JoursFeries.check_zone("Mayotte"):
            return date(year, 4, 27)

        if zone == JoursFeries.check_zone("Martinique"):
            return date(year, 5, 22)

        if zone == JoursFeries.check_zone("Guadeloupe"):
            return date(year, 5, 27)

        if zone == JoursFeries.check_zone("Saint-Martin"):
            if year >= 2018:
                return date(year, 5, 28)
            else:
                return date(year, 5, 27)

        if zone == JoursFeries.check_zone("Guyane"):
            return date(year, 6, 10)

        if zone == JoursFeries.check_zone("Saint-Barthélémy"):
            return date(year, 10, 9)

        if zone == JoursFeries.check_zone("La Réunion") and year >= 1981:
            return date(year, 12, 20)

        return None


jours_feries_france/__init__.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from datetime import date, timedelta


class JoursFeries(object):
    ZONES = [
        "Métropole",
        "Alsace-Moselle",
        "Guadeloupe",
        "Guyane",
        "Martinique",
        "Mayotte",
        "Nouvelle-Calédonie",
        "La Réunion",
        "Polynésie Française",
        "Saint-Barthélémy",
        "Saint-Martin",
        "Wallis-et-Futuna",
        "Saint-Pierre-et-Miquelon",
    ]

    def __init__(self):
        super(JoursFeries, self).__init__()

    @staticmethod
    def check_zone(zone):
        zone = zone or "Métropole"

        if zone not in JoursFeries.ZONES:
            valid_values = ", ".join(JoursFeries.ZONES)
            raise ValueError(
                "%s is invalid. Supported values: %s" % (zone, valid_values)
            )

        return zone

    @staticmethod
    def is_bank_holiday(date, zone=None):
        return date in JoursFeries.for_year(date.year, zone).values()

    @staticmethod
    def next_bank_holiday(date, zone=None):
        while not JoursFeries.is_bank_holiday(date, zone):
            date = date + timedelta(days=1)

        return [
            (k, v)
            for k, v in JoursFeries.for_year(date.year, zone).items()
            if v == date
        ][0]

    @staticmethod
    def for_year(year, zone=None):
        JoursFeries.check_zone(zone)

        bank_holidays = {
            "1er janvier": JoursFeries.premier_janvier(year),
            "1er mai": JoursFeries.premier_mai(year),
            "8 mai": JoursFeries.huit_mai(year),
            "14 juillet": JoursFeries.quatorze_juillet(year),
            "Assomption": JoursFeries.assomption(year),
            "Toussaint": JoursFeries.toussaint(year),
            "11 novembre": JoursFeries.onze_novembre(year),
            "Jour de Noël": JoursFeries.jour_noel(year),
            "Lundi de Pâques": JoursFeries.lundi_paques(year),
            "Ascension": JoursFeries.ascension(year),
            "Lundi de Pentecôte": JoursFeries.lundi_pentecote(year),
            "Vendredi saint": JoursFeries.vendredi_saint(year, zone),
            "2ème jour de Noël": JoursFeries.deuxieme_jour_noel(year, zone),
            "Abolition de l'esclavage": JoursFeries.abolition_esclavage(year, zone),
        }

        bank_holidays = {k: v for k, v in bank_holidays.items() if v}

        return {
            k: v for k, v in sorted(bank_holidays.items(), key=lambda item: item[1])
        }

    @staticmethod
    def paques(year):
        if year < 1886:
            return None
        a = year % 19
        b = year // 100
        c = year % 100
        d = (19 * a + b - b // 4 - ((b - (b + 8) // 25 + 1) // 3) + 15) % 30
        e = (32 + 2 * (b % 4) + 2 * (c // 4) - d - (c % 4)) % 7
        f = d + e - 7 * ((a + 11 * d + 22 * e) // 451) + 114
        month = f // 31
        day = f % 31 + 1
        return date(year, month, day)

    @staticmethod
    def lundi_paques(year):
        if year >= 1886:
            return JoursFeries.paques(year) + timedelta(days=1)
        return None

    @staticmethod
    def vendredi_saint(year, zone):
        if zone == JoursFeries.check_zone("Alsace-Moselle"):
            return JoursFeries.paques(year) - timedelta(days=2)
        return None

    @staticmethod
    def ascension(year):
        if year >= 1802:
            return JoursFeries.paques(year) + timedelta(days=39)
        return None

    @staticmethod
    def lundi_pentecote(year):
        if year >= 1886:
            return JoursFeries.paques(year) + timedelta(days=50)
        return None

    @staticmethod
    def premier_janvier(year):
        if year > 1810:
            return date(year, 1, 1)
        return None

    @staticmethod
    def premier_mai(year):
        if year > 1919:
            return date(year, 5, 1)
        return None

    @staticmethod
    def huit_mai(year):
        if (1953 <= year <= 1959) or year > 1981:
            return date(year, 5, 8)
        return None

    @staticmethod
    def quatorze_juillet(year):
        if year >= 1880:
            return date(year, 7, 14)
        return None

    @staticmethod
    def toussaint(year):
        if year >= 1802:
            return date(year, 11, 1)
        return None

    @staticmethod
    def assomption(year):
        if year >= 1802:
            return date(year, 8, 15)
        return None

    @staticmethod
    def onze_novembre(year):
        if year >= 1918:
            return date(year, 11, 11)
        return None

    @staticmethod
    def jour_noel(year):
        if year >= 1802:
            return date(year, 12, 25)
        return None

    @staticmethod
    def deuxieme_jour_noel(year, zone):
        if zone == JoursFeries.check_zone("Alsace-Moselle"):
            return date(year, 12, 26)
        return None

    @staticmethod
    def abolition_esclavage(year, zone):
        if zone == JoursFeries.check_zone("Mayotte"):
            return date(year, 4, 27)

        if zone == JoursFeries.check_zone("Martinique"):
            return date(year, 5, 22)

        if zone == JoursFeries.check_zone("Guadeloupe"):
            return date(year, 5, 27)

        if zone == JoursFeries.check_zone("Saint-Martin"):
            if year >= 2018:
                return date(year, 5, 28)
            else:
                return date(year, 5, 27)

        if zone == JoursFeries.check_zone("Guyane"):
            return date(year, 6, 10)

        if zone == JoursFeries.check_zone("Saint-Barthélémy"):
            return date(year, 10, 9)

        if zone == JoursFeries.check_zone("La Réunion") and year >= 1981:
            return date(year, 12, 20)

        return None
